### ISSUE TITLE
How to convolve two tensor according to the batch axis?

### ISSUE THREAD

**Issue by bugzhu** [tensorflow/tensorflow#34571]
In my problem, I want to convolve two tensors in my neural network model. The shape of two tensors is [None, 2, 1], [None, 3, 1] respectively. The axis with dimension None means the batch size of the input tensor. For each sample in batch, I want to convolve the two tensors with shape [2, 1] and [3, 1].
However, the tf.nn.conv1d in TensorFlow can only convolve the input with a fixed kernel. Is there any function that can support the convolution of two tensors according to the batch size axis, similar to the tf.multiply which can element wisely multiply two tensors.

**Comment by sahilinneurospace** [comment]
Hi, did you find the solution to this? I'm trying to figure out the same thing.

**Comment by bugzhu** [comment]
I am trying to achiever this, but have no idea till now. Maybe there are some possible solutions in the stackoverflow.  @sahilinneurospace

**Comment by djordjened92** [comment]
Hi @tinyxuyan , I've done some workaround using mentioned tf.nn.conv1d for the same purposes you mentioned in the post:

```
def cross_correlation(vectors, batch_size):
    '''
    The goal is to convolute outputs from both networks,
    each one from the batch of the first network output
    over appropriate one from another network output
    '''
    
    # x and y have shape: [batch_size, vector_size]
    x, y = vectors

    # We need to add 'channels' dimension -> [batch_size, vector_size, 1]
    x = tf.expand_dims(x, -1)
    y = tf.expand_dims(y, -1)
    
    # Firstly we do convolution as the second network
    # output vectors from the batch are all filters.
    # Technically, we've done convolution of all y over each x.
    # From documentation(https://www.tensorflow.org/api_docs/python/tf/nn/conv1d) desired shapes should be:
    # x shape: [batch_shape(batch_size), in_width(vector_size), in_channels(1)]
    # y shape: [filter_width(vector_size), in_channels(1), out_channels(batch_size)]
    # Output shape is [batch_size, vector_length, batch_size].
    conv1d = tf.nn.conv1d(x, tf.transpose(y, perm=[1, 2, 0]), stride=1, padding='SAME')
    conv1d_shape = conv1d.shape
    
    # We need now to apply mask to the output in order to get desired
    # convolution results in the shape [batch_size, vector_length]
    bool_diag = tf.linalg.diag(tf.constant([True] * batch_size))
    mask = tf.repeat(tf.expand_dims(bool_diag, axis=1), conv1d_shape[1], axis=1)
    out = tf.squeeze(tf.ragged.boolean_mask(conv1d, mask), axis=-1)
    
    return tf.reduce_max(out, 1)
```
I hope this is helpful.
